library/pkparse.c

```diff
@@ -487,6 +486,62 @@ static int pk_use_ecparams(const mbedtls_asn1_buf *params, mbedtls_ecp_group *gr
     return 0;
 }
 
+#if defined(MBEDTLS_ECP_DP_CURVE25519_ENABLED) || defined(MBEDTLS_ECP_DP_CURVE448_ENABLED)
+/*
+ * Load an RFC8410 EC key, which doesn't have any parameters
+ */
+static int pk_use_ecparams_curdle(const mbedtls_asn1_buf *params,
+                                  mbedtls_ecp_group_id grp_id,
+                                  mbedtls_ecp_group *grp)
+{
+    if (params->tag != 0 || params->len != 0) {
+        return MBEDTLS_ERR_PK_KEY_INVALID_FORMAT;
+    }
+
+    return mbedtls_ecp_group_load(grp, grp_id);
+}
+
+/*
+ * Parse an RFC 8410 encoded private EC key
+ *
+ * CurvePrivateKey ::= OCTET STRING
+ */
+static int pk_parse_key_curdle_der(mbedtls_ecp_keypair *eck,
+                                   unsigned char *key, size_t keylen, const unsigned char *end,
+                                   int (*f_rng)(void *, unsigned char *, size_t), void *p_rng)
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    size_t len;
+
+    if ((ret = mbedtls_asn1_get_tag(&key, (key + keylen), &len, MBEDTLS_ASN1_OCTET_STRING)) != 0) {
+        return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_PK_KEY_INVALID_FORMAT, ret);
+    }
+
+    if (key + len != end) {
+        return MBEDTLS_ERR_PK_KEY_INVALID_FORMAT;
+    }
+
+    if ((ret = mbedtls_mpi_read_binary_le(&eck->d, key, len)) != 0) {
+        mbedtls_ecp_keypair_free(eck);
+        return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_PK_KEY_INVALID_FORMAT, ret);
+    }
+
+    // reconstruct public key
+    if ((ret = mbedtls_ecp_mul(&eck->grp, &eck->Q, &eck->d, &eck->grp.G,
```

Comment:
Note: conflicts with some of the work done in the meantime, namely https://github.com/Mbed-TLS/mbedtls/pull/7392. For now this is not a problem, but once #7410 is merged this will be: the CI will fail on the merge result of #7410 and this one.

Perhaps this can be fixed by rebasing the present PR on top of recent development, then using `pk_derive_public_key()` here?

Cc @valeriosetti FYI

Reply:
Refactored as proposed.

Is there no way to leverage PSA's math directly? I was able to avoid the cost of one serialize in my refactor, but here you're still doing an extra deserialize privkey+serialize pubkey+deserialize pubkey every time you parse a privkey. This seems really rather inefficient.

Reply:
PSA interfaces are deliberately opaque because they can be implemented by hardware accelerators or in isolated environments (secure element, smartcard, …) that don't expose their internal math.

We're looking at adding PSA interfaces to import/export formatted keys. But those new interfaces would likely directly expose the RFC8410 format, replacing all the `pk*.c` code, so it's unlikely that we'd add a new PSA interface that would help _inside_ pk.